### PR TITLE
feat(fabric): :sparkles: implemented fabric interfaces for iOS and removed buildDir for Android fabric codegen path

### PR DESCRIPTION
**Pull request by yungblud**
## Summary
- created interfaces for connecting with Fabric iOS
- removed buildDir for Android fabric codegen path

### Motivation
- trying to implement Fabric on iOS and Android

### Changes
- created references and files for iOS Fabric header interface files and obc++ files
- removed buildDir build.gradle
- added codegenConfig package.json, modified video specs

## Test plan
1.
Add this `codegenConfig` in package.json (root dir)
```json
"codegenConfig": {
        "name": "RNCVideo",
        "type": "components",
        "jsSrcsDir": "./src/specs"
}
```

2.
Copy all this code fragments in `src/specs/VideoNativeComponent.ts`
<details>
  <summary>
    Show Code
  </summary>
  
```typescript
/* eslint-disable @typescript-eslint/ban-types */
import type {HostComponent, ViewProps} from 'react-native';
import {NativeModules} from 'react-native';
import type {
  DirectEventHandler,
  Double,
  Float,
  Int32,
  WithDefault,
} from 'react-native/Libraries/Types/CodegenTypes';
import codegenNativeComponent from 'react-native/Libraries/Utilities/codegenNativeComponent';

// -------- There are types for native component (future codegen) --------
// if you are looking for types for react component, see src/types/video.ts

type Headers = ReadonlyArray<
  Readonly<{
    key: string;
    value: string;
  }>
>;

type VideoMetadata = Readonly<{
  title?: string;
  subtitle?: string;
  description?: string;
  imageUri?: string;
}>;

export type VideoSrc = Readonly<{
  uri?: string;
  isNetwork?: boolean;
  isAsset?: boolean;
  shouldCache?: boolean;
  type?: string;
  mainVer?: Int32;
  patchVer?: Int32;
  requestHeaders?: Headers;
  startPosition?: Float;
  cropStart?: Float;
  cropEnd?: Float;
  metadata?: VideoMetadata;
  textTracksAllowChunklessPreparation?: boolean; // android
}>;

type DRMType = WithDefault<string, 'widevine'>;

type DebugConfig = Readonly<{
  enable?: boolean;
  thread?: boolean;
}>;

type Drm = Readonly<{
  type?: DRMType;
  licenseServer?: string;
  headers?: Headers;
  contentId?: string; // ios
  certificateUrl?: string; // ios
  base64Certificate?: boolean; // ios default: false
  useExternalGetLicense?: boolean; // ios
}>;

type TextTracks = ReadonlyArray<
  Readonly<{
    title: string;
    language: string;
    type: string;
    uri: string;
  }>
>;

type SelectedTextTrackType = WithDefault<string, 'system'>;

type SelectedAudioTrackType = WithDefault<string, 'system'>;

type SelectedTextTrack = Readonly<{
  type?: SelectedTextTrackType;
  value?: string;
}>;

type SelectedAudioTrack = Readonly<{
  type?: SelectedAudioTrackType;
  value?: string;
}>;

type SelectedVideoTrackType = WithDefault<string, 'auto'>;

type SelectedVideoTrack = Readonly<{
  type?: SelectedVideoTrackType;
  value?: string;
}>;

export type Seek = Readonly<{
  time: Float;
  tolerance?: Float;
}>;

type BufferConfigLive = Readonly<{
  maxPlaybackSpeed?: Float;
  minPlaybackSpeed?: Float;
  maxOffsetMs?: Int32;
  minOffsetMs?: Int32;
  targetOffsetMs?: Int32;
}>;

type BufferingStrategyType = WithDefault<string, 'Default'>;

type BufferConfig = Readonly<{
  minBufferMs?: Float;
  maxBufferMs?: Float;
  bufferForPlaybackMs?: Float;
  bufferForPlaybackAfterRebufferMs?: Float;
  maxHeapAllocationPercent?: Float;
  backBufferDurationMs?: Float; // Android
  minBackBufferMemoryReservePercent?: Float;
  minBufferMemoryReservePercent?: Float;
  cacheSizeMB?: Float;
  live?: BufferConfigLive;
}>;

type SubtitleStyle = Readonly<{
  fontSize?: Float;
  paddingTop?: WithDefault<Float, 0>;
  paddingBottom?: WithDefault<Float, 0>;
  paddingLeft?: WithDefault<Float, 0>;
  paddingRight?: WithDefault<Float, 0>;
  opacity?: WithDefault<Float, 1>;
}>;

type OnLoadData = Readonly<{
  currentTime: Float;
  duration: Float;
  naturalSize: Readonly<{
    width: Float;
    height: Float;
    orientation: WithDefault<string, 'landscape'>;
  }>;
  audioTracks: {
    index: Int32;
    title?: string;
    language?: string;
    bitrate?: Float;
    type?: string;
    selected?: boolean;
  }[];
  textTracks: {
    index: Int32;
    title?: string;
    language?: string;
    /**
     * iOS only supports VTT, Android supports all 3
     */
    type?: WithDefault<string, 'srt'>;
    selected?: boolean;
  }[];
}>;

export type OnLoadStartData = Readonly<{
  isNetwork: boolean;
  type: string;
  uri: string;
}>;

export type OnVideoAspectRatioData = Readonly<{
  width: Float;
  height: Float;
}>;

export type OnBufferData = Readonly<{isBuffering: boolean}>;

export type OnProgressData = Readonly<{
  currentTime: Float;
  playableDuration: Float;
  seekableDuration: Float;
}>;

export type OnBandwidthUpdateData = Readonly<{
  bitrate: Int32;
  width?: Float;
  height?: Float;
  trackId?: Int32;
}>;

export type OnSeekData = Readonly<{
  currentTime: Float;
  seekTime: Float;
}>;

export type OnPlaybackStateChangedData = Readonly<{
  isPlaying: boolean;
}>;

export type OnTimedMetadataData = Readonly<{
  metadata: {
    value?: string;
    identifier: string;
  }[];
}>;

export type OnAudioTracksData = Readonly<{
  audioTracks: {
    index: Int32;
    title?: string;
    language?: string;
    bitrate?: Float;
    type?: string;
    selected?: boolean;
  }[];
}>;

type OnTextTracksData = Readonly<{
  textTracks: {
    index: Int32;
    title?: string;
    language?: string;
    /**
     * iOS only supports VTT, Android supports all 3
     */
    type?: WithDefault<string, 'srt'>;
    selected?: boolean;
  }[];
}>;

export type OnTextTrackDataChangedData = Readonly<{
  subtitleTracks: string;
}>;

export type OnVideoTracksData = Readonly<{
  videoTracks: {
    index: Int32;
    tracksId?: string;
    codecs?: string;
    width?: Float;
    height?: Float;
    bitrate?: Float;
    selected?: boolean;
  }[];
}>;

export type OnPlaybackRateChangeData = Readonly<{
  playbackRate: Float;
}>;

export type OnVolumeChangeData = Readonly<{
  volume: Float;
}>;

export type OnExternalPlaybackChangeData = Readonly<{
  isExternalPlaybackActive: boolean;
}>;

export type OnGetLicenseData = Readonly<{
  licenseUrl: string;
  loadedLicenseUrl: string;
  contentId: string;
  spcBase64: string;
}>;

export type OnPictureInPictureStatusChangedData = Readonly<{
  isActive: boolean;
}>;

type OnReceiveAdEventData = Readonly<{
  data?: {};
  event: WithDefault<string, 'AD_BREAK_ENDED'>;
}>;

export type OnVideoErrorData = Readonly<{
  error: Readonly<{
    errorString?: string; // android
    errorException?: string; // android
    errorStackTrace?: string; // android
    errorCode?: string; // android
    error?: string; // ios
    code?: Int32; // ios
    localizedDescription?: string; // ios
    localizedFailureReason?: string; // ios
    localizedRecoverySuggestion?: string; // ios
    domain?: string; // ios
  }>;
  target?: Int32; // ios
}>;

export type OnAudioFocusChangedData = Readonly<{
  hasAudioFocus: boolean;
}>;

type ControlsStyles = Readonly<{
  hideSeekBar?: boolean;
  seekIncrementMS?: Float;
}>;

export type OnControlsVisibilityChange = Readonly<{
  isVisible: boolean;
}>;

export interface VideoNativeProps extends ViewProps {
  src?: VideoSrc;
  drm?: Drm;
  adTagUrl?: string;
  allowsExternalPlayback?: boolean; // ios, true
  maxBitRate?: Float;
  resizeMode?: WithDefault<string, 'none'>;
  repeat?: boolean;
  automaticallyWaitsToMinimizeStalling?: boolean;
  textTracks?: TextTracks;
  selectedTextTrack?: SelectedTextTrack;
  selectedAudioTrack?: SelectedAudioTrack;
  selectedVideoTrack?: SelectedVideoTrack; // android
  paused?: boolean;
  muted?: boolean;
  controls?: boolean;
  filter?: WithDefault<string, ''>;
  filterEnabled?: boolean;
  volume?: Float; // default 1.0
  playInBackground?: boolean;
  preventsDisplaySleepDuringVideoPlayback?: boolean;
  preferredForwardBufferDuration?: Float; //ios, 0
  playWhenInactive?: boolean; // ios, false
  pictureInPicture?: boolean; // ios, false
  ignoreSilentSwitch?: WithDefault<string, 'inherit'>; // ios, 'inherit'
  mixWithOthers?: WithDefault<string, 'inherit'>; // ios, 'inherit'
  rate?: Float;
  fullscreen?: boolean; // ios, false
  fullscreenAutorotate?: boolean;
  fullscreenOrientation?: WithDefault<string, 'all'>;
  progressUpdateInterval?: Float;
  restoreUserInterfaceForPIPStopCompletionHandler?: boolean;
  localSourceEncryptionKeyScheme?: string;
  debug?: DebugConfig;
  showNotificationControls?: WithDefault<boolean, false>; // Android, iOS
  bufferConfig?: BufferConfig; // Android
  contentStartTime?: Int32; // Android
  currentPlaybackTime?: Double; // Android
  disableDisconnectError?: boolean; // Android
  focusable?: boolean; // Android
  hideShutterView?: boolean; //	Android
  minLoadRetryCount?: Int32; // Android
  reportBandwidth?: boolean; //Android
  subtitleStyle?: SubtitleStyle; // android
  viewType?: Int32; // Android
  bufferingStrategy?: BufferingStrategyType; // Android
  controlsStyles?: ControlsStyles; // Android
  onControlsVisibilityChange?: DirectEventHandler<OnControlsVisibilityChange>;
  onVideoLoad?: DirectEventHandler<OnLoadData>;
  onVideoLoadStart?: DirectEventHandler<OnLoadStartData>;
  onVideoAspectRatio?: DirectEventHandler<OnVideoAspectRatioData>;
  onVideoBuffer?: DirectEventHandler<OnBufferData>;
  onVideoError?: DirectEventHandler<OnVideoErrorData>;
  onVideoProgress?: DirectEventHandler<OnProgressData>;
  onVideoBandwidthUpdate?: DirectEventHandler<OnBandwidthUpdateData>;
  onVideoSeek?: DirectEventHandler<OnSeekData>;
  onVideoEnd?: DirectEventHandler<{}>; // all
  onVideoAudioBecomingNoisy?: DirectEventHandler<{}>;
  onVideoFullscreenPlayerWillPresent?: DirectEventHandler<{}>; // ios, android
  onVideoFullscreenPlayerDidPresent?: DirectEventHandler<{}>; // ios, android
  onVideoFullscreenPlayerWillDismiss?: DirectEventHandler<{}>; // ios, android
  onVideoFullscreenPlayerDidDismiss?: DirectEventHandler<{}>; // ios, android
  onReadyForDisplay?: DirectEventHandler<{}>;
  onPlaybackRateChange?: DirectEventHandler<OnPlaybackRateChangeData>; // all
  onVolumeChange?: DirectEventHandler<OnVolumeChangeData>; // android, ios
  onVideoExternalPlaybackChange?: DirectEventHandler<OnExternalPlaybackChangeData>;
  onGetLicense?: DirectEventHandler<OnGetLicenseData>;
  onPictureInPictureStatusChanged?: DirectEventHandler<OnPictureInPictureStatusChangedData>;
  onRestoreUserInterfaceForPictureInPictureStop?: DirectEventHandler<{}>;
  onReceiveAdEvent?: DirectEventHandler<OnReceiveAdEventData>;
  onVideoPlaybackStateChanged?: DirectEventHandler<OnPlaybackStateChangedData>; // android only
  onVideoIdle?: DirectEventHandler<{}>; // android only (nowhere in document, so do not use as props. just type declaration)
  onAudioFocusChanged?: DirectEventHandler<OnAudioFocusChangedData>; // android only (nowhere in document, so do not use as props. just type declaration)
  onTimedMetadata?: DirectEventHandler<OnTimedMetadataData>; // ios, android
  onAudioTracks?: DirectEventHandler<OnAudioTracksData>; // android
  onTextTracks?: DirectEventHandler<OnTextTracksData>; // android
  onTextTrackDataChanged?: DirectEventHandler<OnTextTrackDataChangedData>; // iOS
  onVideoTracks?: DirectEventHandler<OnVideoTracksData>; // android
}

export type VideoComponentType = HostComponent<VideoNativeProps>;

export type VideoSaveData = {
  uri: string;
};

export interface VideoManagerType {
  save: (option: object, reactTag: number) => Promise<VideoSaveData>;
  seek: (option: Seek, reactTag: number) => Promise<void>;
  setPlayerPauseState: (paused: boolean, reactTag: number) => Promise<void>;
  setLicenseResult: (
    result: string,
    licenseUrl: string,
    reactTag: number,
  ) => Promise<void>;
  setLicenseResultError: (
    error: string,
    licenseUrl: string,
    reactTag: number,
  ) => Promise<void>;
  setVolume: (volume: number, reactTag: number) => Promise<void>;
  getCurrentPosition: (reactTag: number) => Promise<number>;
  setFullScreen: (fullScreen: boolean, reactTag: number) => Promise<void>;
}

export interface VideoDecoderPropertiesType {
  getWidevineLevel: () => Promise<number>;
  isCodecSupported: (
    mimeType: string,
    width: number,
    height: number,
  ) => Promise<'unsupported' | 'hardware' | 'software'>;
  isHEVCSupported: () => Promise<'unsupported' | 'hardware' | 'software'>;
}

export const VideoManager = NativeModules.VideoManager as VideoManagerType;
export const VideoDecoderProperties =
  NativeModules.VideoDecoderProperties as VideoDecoderPropertiesType;

export default codegenNativeComponent<VideoNativeProps>(
  'RNCVideo',
) as VideoComponentType;

```

</details>

3.
run codegen in root directory
`yarn codegen`

4.
`cd examples/FabricExample/ios`

`bundle exec pod install`

`open FabricExample.xcworkspace`

and run build from XCode (FabricExample xcworkspace)

edited by @KrzysztofMoch 2.07.2024

### CODE REVIEW

File: ios/Fabric/Video/RCTVideo.h [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+//- (void)onRestoreUserInterfaceForPictureInPictureStop;
+//- (void)onPictureInPictureStatusChangedWithIsActive:(BOOL)isActive;
+//- (void)onPlaybackRateChangeWithPlaybackRate:(NSNumber *)playbackRate;
+//- (void)onVideoExternalPlaybackChangeWithIsExternalPlaybackActive:(BOOL)isExternalPlaybackActive;
+//- (void)onReceiveAdEventWithEvent:(NSString *)event;
+
+@end
+
+#if __has_include(<react-native-video/RCTVideoCache.h>)
+@interface RCTVideo : UIView <RCTVideoPlayerViewControllerDelegate, DVAssetLoaderDelegatesDelegate, AVAssetResourceLoaderDelegate>
```
Comment (KrzysztofMoch) [inline ios/Fabric/Video/RCTVideo.h]: Shouldn't this one also contain `AVPictureInPictureControllerDelegate` ?

**Comment by freeboub** [comment]
There are a lot of commented code in this PR :/

**Comment by freeboub** [comment]
@yungblud are you on discord ? can you please ping me ?

**Comment by yungblud** [comment]
> There are a lot of commented code in this PR :/ 

Yep, but it can be easily removed if you want.

**Comment by yungblud** [comment]
> @yungblud are you on discord ? can you please ping me ?

@freeboub okay. Add me with "killi8n"

**Comment by yungblud** [comment]
> There are a lot of commented code in this PR :/

@freeboub First, that comments are not meaningless. But I can simplify comments!

File: ios/Fabric/Video/RCTVideo.h (hunk 1 above)
Comment (yungblud) [reply ios/Fabric/Video/RCTVideo.h:45]: Yes I can find out `AVPictureInPictureControllerDelegate` in RCTPictureInPicture.swift
I will add it!

**Comment by yungblud** [comment]
> @yungblud are you on discord ? can you please ping me ?

@freeboub What is your discord id?
If you have anything to ask about this PR, we can talk about them in this PR!

File: ios/Fabric/Video/RCTVideo.h (hunk 1 above)
Comment (yungblud) [reply ios/Fabric/Video/RCTVideo.h]: @KrzysztofMoch Hmm, I am not sure.
That RCTVideo interface is for `#if __has_include(<react-native-video/RCTVideoCache.h>)`
Do we need to extends `AVPictureInPictureControllerDelegate` for RCTVideoCache too?

There is
```objc
#else
@interface RCTVideo : UIView <RCTVideoPlayerViewControllerDelegate, AVPictureInPictureControllerDelegate, AVAssetResourceLoaderDelegate>
#endif
```
which has extended `AVPictureInPictureControllerDelegate` in the last endif.

**Comment by yungblud** [comment]
@KrzysztofMoch @freeboub I think we should focus on interfaces more in this PR rather than focusing any other things.
Implementations are not main purpose for this PR.
Somebody has to create the base of Fabric. I am doing that!

**Comment by KrzysztofMoch** [comment]
@yungblud give us a moment - I think we will create `new-arch` branch that we will keep sync with master but we will merge step by step new arch changes as partial migration wont work in react-native. I need to discuss this internally

**Comment by yungblud** [comment]
> @yungblud give us a moment - I think we will create `new-arch` branch that we will keep sync with master but we will merge step by step new arch changes as partial migration wont work in react-native. I need to discuss this internally

@KrzysztofMoch Okay, that sounds great.
I am stacking some Fabric features in my forked repository. Maybe that would be synced by your roadmaps.

**Comment by freeboub** [comment]
@yungblud I sent you an invitation, my point is not directly linked to this PR, more a matter of organization (I don't want to stack your PRs for long time ...)

File: src/specs/VideoNativeComponent.ts [hunk 2]
```diff
@@ -286,7 +287,7 @@ export type OnAudioFocusChangedData = Readonly<{
 
 type ControlsStyles = Readonly<{
   hideSeekBar?: boolean;
-  seekIncrementMS?: number;
+  seekIncrementMS?: Float;
```
Comment (freeboub) [inline src/specs/VideoNativeComponent.ts]: an Int32 shall be enough here ?!

Comment (yungblud) [reply src/specs/VideoNativeComponent.ts]: Yep, I think it would be better!
if we don't use dots with numbers

**Comment by yungblud** [comment]
@KrzysztofMoch @freeboub changed base branch to `6.5-dev`!

File: ios/Fabric/Video/RCTVideo.h (hunk 1 above)
Comment (KrzysztofMoch) [reply ios/Fabric/Video/RCTVideo.h]: I think we should - otherwise we won't be able to use PIP when cache is enabled (I think)

**Comment by KrzysztofMoch** [comment]
@yungblud Swift Formatter is failing  you need to run `yarn check-ios` and it should be fine

File: ios/Fabric/Video/RCTVideo.h (hunk 1 above)
Comment (yungblud) [reply ios/Fabric/Video/RCTVideo.h]: Okay.
added extend AVPictureInPictureControllerDelegate when cache is enabled!
[feat: 🎨 extends AVPictureInPictureControllerDelegate when cache is en…](https://github.com/TheWidlarzGroup/react-native-video/pull/3965/commits/5fa3c095d1834a73125448a937cf74fd54857b28)

**Comment by yungblud** [comment]
> @yungblud Swift Formatter is failing you need to run `yarn check-ios` and it should be fine

@KrzysztofMoch Passed all github CI checks!

**Pull request closed without merging**